wgengine/netstack/netstack.go

```diff
@@ -381,18 +382,71 @@ func (ns *Impl) acceptUDP(r *udp.ForwarderRequest) {
 		ns.logf("Could not create endpoint, exiting")
 		return
 	}
+	localAddr, err := ep.GetLocalAddress()
+	if err != nil {
+		return
+	}
+	remoteAddr, err := ep.GetRemoteAddress()
+	if err != nil {
+		return
+	}
 	c := gonet.NewUDPConn(ns.ipstack, &wq, ep)
-	go echoUDP(c)
+	go ns.forwardUDP(c, &wq, localAddr, remoteAddr)
 }
 
-func echoUDP(c *gonet.UDPConn) {
-	buf := make([]byte, 1500)
-	for {
-		n, err := c.Read(buf)
+func (ns *Impl) forwardUDP(client *gonet.UDPConn, wq *waiter.Queue, clientLocalAddr, clientRemoteAddr tcpip.FullAddress) {
+	port := clientLocalAddr.Port
+	ns.logf("[v2] netstack: forwarding incoming UDP connection on port %v", port)
+	serverLocalAddr := &net.UDPAddr{Port: int(clientRemoteAddr.Port)}
+	serverRemoteAddr := &net.UDPAddr{IP: net.ParseIP("127.0.0.1"), Port: int(port)}
+	server, err := net.ListenUDP("udp", serverLocalAddr)
+	if err != nil {
+		ns.logf("netstack: could not bind local port %v: %v, trying again with random port", clientRemoteAddr.Port, err)
+		server, err = net.ListenUDP("udp", nil)
 		if err != nil {
-			break
+			ns.logf("netstack: could not connect to local UDP server on port %v: %v", port, err)
+			return
 		}
-		c.Write(buf[:n])
 	}
-	c.Close()
+	var cancels [2]context.CancelFunc
+	timer := time.AfterFunc(2*time.Minute, func() {
+		for _, c := range cancels {
+			if c != nil {
+				c()
+			}
+		}
+		ns.logf("netstack: forwarder UDP connection on port %v closed", port)
+	})
+	cancels[0] = startPacketCopy(client, &net.UDPAddr{
+		IP:   net.ParseIP(clientRemoteAddr.Addr.String()),
+		Port: int(clientRemoteAddr.Port),
+	}, server, ns.logf, timer)
+	cancels[1] = startPacketCopy(server, serverRemoteAddr, client, ns.logf, timer)
+
+}
+
+func startPacketCopy(dst net.PacketConn, dstAddr net.Addr, src net.PacketConn, logf logger.Logf, timer *time.Timer) context.CancelFunc {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				break
+			default:
+				pkt := make([]byte, mtu)
+				n, srcAddr, err := src.ReadFrom(pkt)
+				if err != nil {
+					logf("read packet from %s failed: %v", srcAddr, err)
+					continue
+				}
+				_, err = dst.WriteTo(pkt[:n], dstAddr)
+				if err != nil {
+					logf("write packet to %s failed: %v", dstAddr, err)
+				}
+				logf("wrote UDP packet %s -> %s", srcAddr, dstAddr)
+				timer.Reset(2 * time.Minute)
```

Comment:
maybe rather than passing a timer and hard-coding 2 minutes here, pass an `extend func()` and put the `timer.Reset(2*time.Minute)` in the caller when creating the closure to pass down as `extend`